### DISCUSSION TITLE
Optimizing data projection queries ($select)

### DISCUSSION THREAD

**Discussion by swurzinger** [AutoMapper/AutoMapper.Extensions.OData#228] [Q&A] [answered]
An OData query with a simple projection like `/odata/orders?$select=Id` seems to ignore the fact that only the Id fields would be necessary in this case and still fetches all fields defined in the entity:
```SQL
SELECT [o].[Id], [o].[OrderDate], [o].[OrderType], [o].[Weight], [o].[MaterialValue], [o].[Currency], [o].[OrderStatus]
FROM [order].[Order] AS [o]
ORDER BY [o].[Id]
```
What I'd expect would be a SQL query like this:
```SQL
SELECT [o].[Id]
FROM [order].[Order] AS [o]
ORDER BY [o].[Id]
```
This is especially hurts performance when additional tables are joined.

The OData controller implementation is as simple as:
```csharp
public async Task<ActionResult<IQueryable<Order>>> GetAsync(ODataQueryOptions<Order> options)
{
	var results = await _dbContext.Orders.GetQueryAsync<Dto.Order, Entity.Order>(_mapper, options);
	return Ok(results);
}
```

Is there an issue somewhere in the implementation/configuration or is this just not supported?

**Comment by BlaiseD** [comment, marked as answer]
The [tests](https://github.com/AutoMapper/AutoMapper.Extensions.OData/blob/6a68bac6af428e34bb3e60fb8d43f4fcd1ead469/AutoMapper.OData.EFCore.Tests/GetQuerySelectTests.cs#L33) show that `$select` works.  Maybe you're missing explicit expansion in your configuration.  See the [ReadMe](https://github.com/AutoMapper/AutoMapper.Extensions.OData/blob/master/README.md).

**Reply by swurzinger** [reply]
Thank you. It works indeed when enabling explicit expansion for all fields. I enabled it only for certain fields (navigation properties) because in my case that causes issues with some fields, which aren't correctly expanded then. I think it's the same issue as described in https://github.com/AutoMapper/AutoMapper.Extensions.OData/issues/196 .

I have fields like this:
```csharp
public record Order(Guid Id, List<OrderRemark> Remarks, ...);
public record OrderRemark(string IsoLanguageCode, string Text);
```

The `Id` of `OrderRemark` is internal and not mapped in the Dto, thus `$expand` can't be used, as it only works for navigation properties (and navigation properties require a key).
If I enable explicit expansion for `Remarks` the list will only be populated when explicitly requested in `$select` (#196).
If I disable explicit expansion for `Remarks` the order remarks table will always be joined, even if I just select order `Id` (#228).

**Reply by BlaiseD** [reply]
Yep - it's a known issue.  Today we have a hardcoded list of types which we expand by default (including lists of those types).  We need a PR to expand all non-navigation [properties by default](https://github.com/BpsLogicBuilder/LogicBuilder.DataComponents/pull/55).

**Reply by BlaiseD** [reply]
Should be resolved now with PR #229 .
